### ISSUE TITLE
Azure Function doesn't accept Key Vault Reference for Storage Connection String for consumption plan

### ISSUE THREAD

**Issue by callppatel** [hashicorp/terraform-provider-azurerm#7434]
When we creating the Azure Function by referring Storage connection String as "Key Vault" String, we are getting below error message. Thru UI we can update connection string with Key Vault Reference but thru Terraform we can't.

Error message
Error: web.AppsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="BadRequest" Message="The parameter 'WEBSITE_CONTENTAZUREFILECONNECTIONSTRING' has an invalid value. Details: Cannot specify key vault references not referencing User Assigned Identity on Create Site." Details=[{"Message":"The parameter 'WEBSITE_CONTENTAZUREFILECONNECTIONSTRING' has an invalid value. Details: Cannot specify key vault references not referencing User Assigned Identity on Create Site."},{"Code":"BadRequest"},{"ErrorEntity":{"Code":"BadRequest","ExtendedCode":"01033","Message":"The parameter 'WEBSITE_CONTENTAZUREFILECONNECTIONSTRING' has an invalid value. Details: Cannot specify key vault references not referencing User Assigned Identity on Create Site.","MessageTemplate":"The parameter '{0}' has an invalid value. Details: {1}.","Parameters":["WEBSITE_CONTENTAZUREFILECONNECTIONSTRING","Cannot specify key vault references not referencing User Assigned Identity on Create Site"]}}]

This could be related to below issues:
 https://github.com/microsoft/vscode-azurefunctions/issues/1934
 https://github.com/Azure/azure-functions-host/issues/5306

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version
Terraform v0.12.23
+ provider.archive v1.3.0
+ provider.azurerm v2.5.0

### Affected Resource(s)

* `azurerm_function_app`

### Terraform Configuration Files
storage_connection_string = "@Microsoft.KeyVault(SecretUri=${azurerm_key_vault_secret.this.id})"

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://keybase.io/hashicorp
```

### Debug Output

Error: web.AppsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="BadRequest" Message="The parameter 'WEBSITE_CONTENTAZUREFILECONNECTIONSTRING' has an invalid value. Details: Cannot specify key vault references not referencing User Assigned Identity on Create Site." Details=[{"Message":"The parameter 'WEBSITE_CONTENTAZUREFILECONNECTIONSTRING' has an invalid value. Details: Cannot specify key vault references not referencing User Assigned Identity on Create Site."},{"Code":"BadRequest"},{"ErrorEntity":{"Code":"BadRequest","ExtendedCode":"01033","Message":"The parameter 'WEBSITE_CONTENTAZUREFILECONNECTIONSTRING' has an invalid value. Details: Cannot specify key vault references not referencing User Assigned Identity on Create Site.","MessageTemplate":"The parameter '{0}' has an invalid value. Details: {1}.","Parameters":["WEBSITE_CONTENTAZUREFILECONNECTIONSTRING","Cannot specify key vault references not referencing User Assigned Identity on Create Site"]}}]

### Steps to Reproduce

1. `terraform apply`

### References

* #0000

**Comment by neil-yechenwei** [comment]
@callppatel, thanks for opening this issue. Could you provide your tf config that triggers this issue for further investigation? Thanks.

**Comment by callppatel** [comment]
> @callppatel, thanks for opening this issue. Could you provide your tf config that triggers this issue for further investigation? Thanks.


```ruby
locals {
  loc_short_name = {
    "centralus"     = "usc"
    "eastus2"       = "use2"
    "northeurope"   = "eun"
    "westeurope"    = "euw"
    "eastasia"      = "ase"
    "southeastasia" = "asse"
  }

  trimmed_loc = local.loc_short_name[data.azurerm_resource_group.this.location]
  # Name
  default_application_settings = {
    FUNCTIONS_WORKER_RUNTIME = var.function_language
  }
}

# Getting the client config with current Service Principal
data "azurerm_client_config" "current" {}

data "azurerm_resource_group" "this" {
  name = var.resource_group
}

data "azurerm_key_vault" "this" {
  name                = var.keyvault_name
  resource_group_name = data.azurerm_resource_group.this.name
}

resource "azurerm_storage_account" "this" {
  resource_group_name       = data.azurerm_resource_group.this.name
  name                      = "sttestfunc123tionstorga"
  location                  = data.azurerm_resource_group.this.location
  account_tier              = "Standard"
  account_replication_type  = "LRS"
  account_kind              = "StorageV2"
  access_tier               = "Hot"
  is_hns_enabled            = false
  enable_https_traffic_only = true
}

resource "azurerm_app_service_plan" "this" {
  name                = var.function_name
  location            = data.azurerm_resource_group.this.location
  resource_group_name = data.azurerm_resource_group.this.name
  kind                = "FunctionApp"
  sku {
    tier = "Dynamic"
    size = "Y1"
  }
}


resource "azurerm_function_app" "this" {
  name                      = var.function_name
  location                  = data.azurerm_resource_group.this.location
  resource_group_name       = data.azurerm_resource_group.this.name
  app_service_plan_id       = azurerm_app_service_plan.this.id 
  storage_connection_string = "@Microsoft.KeyVault(SecretUri=${azurerm_key_vault_secret.this.id})"
  identity {
    type = "SystemAssigned"
  }

  app_settings = merge(
    local.default_application_settings,
    var.function_app_application_settings,
  )
  site_config {
    always_on = false
  }
  dynamic "connection_string" {
    for_each = var.connection_strings
    content {
      name  = connection_string.value.name
      type  = connection_string.value.type
      value = connection_string.value.value
    }
  }

  auth_settings {
    enabled = true
    active_directory {
      client_id = var.auth_client_id
      allowed_audiences = var.auth_allowed_audiences
    }
    default_provider = "AzureActiveDirectory"
  }
  depends_on = [azurerm_key_vault_secret.this]
}
## Storing the Storage Account Key to Key Vault
resource "azurerm_key_vault_secret" "this" {
  name         = var.function_name
  value        = azurerm_storage_account.this.primary_connection_string
  key_vault_id = data.azurerm_key_vault.this.id
  depends_on   = [azurerm_storage_account.this]
}

resource "azurerm_key_vault_access_policy" "this" {
  count        = 1
  key_vault_id = data.azurerm_key_vault.this.id
  tenant_id    = data.azurerm_client_config.current.tenant_id
  object_id    = lookup(azurerm_function_app.this.identity[0], "principal_id")

  secret_permissions = [
    "get", "list"
  ]
  depends_on = [azurerm_function_app.this]
}

provider "azurerm" {
  version = "=2.5.0"
  features {}
}
```

**Comment by callppatel** [comment]
Let me know if need any info, by looking the link which I mentioned in my original request. It looks like API issue, because I can see the same issue is reported when they do deploy thru vscode as well. 

I am able to update above mention environment variable "WEBSITE_CONTENTAZUREFILECONNECTIONSTRING" with Key Vault Reference thru UI.

**Comment by neil-yechenwei** [comment]
Yes. I assume you're right. I raised an issue on Azure/azure-rest-api-specs#9947

**Comment by callppatel** [comment]
any update on this.

**Comment by petrkoutnycz** [comment]
any updates on this? We'd appreciate this as well. Thank you

**Comment by callppatel** [comment]
This is not resolve yet, but we have workaround, first you let function to be created with without key vault reference. once function is created use Azure CLi command to update the setting like this below

```
resource "null_resource" "update_setting_consumption_plan" {
  provisioner "local-exec" {
    when        = create
    interpreter = ["pwsh", "-command"]
    command     = <<EOT
      sleep 30
      az functionapp config appsettings set --name ${azurerm_function_app.this.name} --resource-group ${var.resource_group} `
      --settings 'WEBSITE_CONTENTAZUREFILECONNECTIONSTRING="@Microsoft.KeyVault(SecretUri=${module.azure_key_vault_secret.secret_identifier})"' `
      'AzureWebJobsStorage="@Microsoft.KeyVault(SecretUri=${module.azure_key_vault_secret.secret_identifier})"' 
    EOT
  }
}
```

**Comment by mishaskalova** [comment]
any ETA on the fix for this issue..?

**Comment by katbyte** [comment]
Thank you for taking the time to report this issue, and apologies that it has sat without a resolution for so long.

This bug was reported against an early major version of the AzureRM provider. The provider has since moved on to v5.x, and the affected functionality has often been significantly reworked across those major releases, so bug reports from legacy major versions prior to v4.0 are generally no longer actionable as written.

As part of a cleanup of historical issues we are closing bug reports that target legacy provider versions. If you are still running into this behaviour on the current major version, please open a new issue with up-to-date reproduction details (feel free to reference this one) so it can be triaged with current context.

Thank you again for the report, and for your patience.
